### PR TITLE
[CH #57710] Interactivity popups

### PR DESCRIPTION
**Pull request by manmorjim**
### Changes

- Fixes issue: sometimes the feature cannot be clicked

- Fixes issue: Cannot disable hovers

- Add cursor pointer when the mouse is on a feature

- Add Airship popup component as Popup style

- Improves examples: Examples similar to VL: https://carto.com/developers/carto-vl/examples/#example-add-pop-ups-on-click
  - Feature hover
  - Feature click
  - Add pop-ups on hover
  - Add pop-ups on click
  - Interactive based styling

- Fixes issue: event error if layer was not added to the map

### CODE REVIEW

File: examples/visualization/interactions/add-popups-onclick.html [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    />
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://carto.com/developers/carto-vl/v1.4.4/examples/maps/style.css"
+    />
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://libs.cartocdn.com/airship-style/v2.1.1/airship.css"
```
Comment (VictorVelarde) [inline examples/visualization/interactions/add-popups-onclick.html]: We can point to minor, and get patches for free, if we use a url like:
https://libs.cartocdn.com/airship-style/v2.1/airship.css (it also works with https://libs.cartocdn.com/airship-style/v2/airship.css)

Comment (VictorVelarde) [reply examples/visualization/interactions/add-popups-onclick.html]: ... and the last one is 2.3 (you can use minified versions too, like airship.min.css). In fact, I think I'm gonna add the lib to 'template' for other examples

**Review by alasarr — CHANGES_REQUESTED** [review]
It looks great and it works great! Congrats on the nice work.

As an improvement, I feel that Layer.ts is growing too much. Could you try to move some logic of popup to an external module? 

Something like LayerInteration.ts where you can encapsulate all the code related to the layer and the interactivity, so you can define there the code of layer.on, _setPopupHandler, setPopupHover, setPopupClick, _wrapInteractiveStyle

File: examples/interactivity/add-popups-onclick.html [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    <!-- Custom CSS -->
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://api.tiles.mapbox.com/mapbox-gl-js/v1.10.0/mapbox-gl.css"
+    />
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://carto.com/developers/carto-vl/v1.4.4/examples/maps/style.css"
```
Comment (VictorVelarde) [inline examples/interactivity/add-popups-onclick.html]: We should remove this VL-thing. At the very end, it's equivalent to our map 100% custom style

File: examples/interactivity/add-popups-onclick.html [hunk 3]
```diff
@@ -0,0 +1,104 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>GeoJSON Layer with CARTO Tiles Example</title>
```
Comment (VictorVelarde) [inline examples/interactivity/add-popups-onclick.html]: Not urgent, but we should update titles too

File: examples/interactivity/add-popups-onclick.html [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      #map {
+        width: 100vw;
+        height: 100vh;
+      }
+
+      div.as-infowindow > div.as-infowindow__content > button.as-btn {
+        float: right;
+        margin-top: -10px;
+        margin-right: -10px;
+      }
```
Comment (VictorVelarde) [inline examples/interactivity/add-popups-onclick.html:40]: This is ok for now. I will try to update the styles for examples in the near future, to use airship all around (so styling is homogeneous with popups)

File: examples/interactivity/add-popups-onhover.html [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        const populationLayer = new carto.viz.Layer(
+          'world_population_2015',
+          sizeBinStyle
+        );
+        await populationLayer.setPopupHover([
+          'cartodb_id',
+          { attr: 'pop_2015', title: 'Population' }
+        ]);
+        await populationLayer.addTo(deckMap);
+        window.deckMap = deckMap;
```
Comment (VictorVelarde) [inline examples/interactivity/add-popups-onhover.html]: I would remove the window.deckMap if not required. This is nice for debugging, but we should try to be homogenous on examples. These examples will be a very relevant part of the documentation!

File: packages/viz/src/lib/layer/Layer.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  // Deck.gl Map instance
+  private _deckInstance: Deck | undefined;
+
+  // Instance to the DeckLayer of the instance
+  // It cannot be a reference to (import { Layer } from '@deck.gl/core') because
+  // the typing of getPickinfo method is different from TileLayer and Layer are
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _deckLayer?: any;
+
+  private _interacitivity: LayerInteractivity;
```
Comment (VictorVelarde) [inline packages/viz/src/lib/layer/Layer.ts]: 👀 typo: `_interacitivity`

File: packages/viz/src/lib/layer/Layer.ts [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+      clickStyle = buildStyle(options.clickStyle);
+    }
+
+    this._interacitivity = new LayerInteractivity(
+      this,
+      this.getStyle.bind(this),
+      this.setStyle.bind(this),
+      hoverStyle,
+      clickStyle
+    );
```
Comment (VictorVelarde) [inline packages/viz/src/lib/layer/Layer.ts]: How about a small refactor, having a helper method to do `this._interactivity = buildInteractivity(options)`?

File: packages/viz/src/lib/layer/Layer.ts [hunk 8]
```diff
@@ -22,19 +28,38 @@ export class Layer implements StyledLayer {
...
     this._options = {
       id: `${this._source.id}-${Date.now()}`,
+      ...this._interacitivity.getDefaultOptions(),
```
Comment (VictorVelarde) [inline packages/viz/src/lib/layer/Layer.ts]: typo here

File: packages/viz/src/lib/layer/Layer.ts [hunk 9]
```diff
@@ -201,51 +227,15 @@ export class Layer implements StyledLayer {
...
-      this._options.onClick = clickHandler;
-      this._options.pickable = true;
-    } else {
-      if (this._clickPopup) {
-        this._clickPopup.close();
-      }
-
-      this._options.onClick = undefined;
-    }
+    this._interactivity.setPopupClick(elements);
```
Comment (alasarr) [inline packages/viz/src/lib/layer/Layer.ts:230]: Vamos!! 👌

**Review by alasarr — APPROVED** [review]

**Pull request merged**